### DISCUSSION TITLE
Best Practise auto run on commit

### DISCUSSION THREAD

**Discussion by cdpb** [semaphoreui/semaphore#1103] [Q&A]
Hello everyone,

first, thank you @fiftin and others for creating this amazing piece of software.

The documentation is a bit sparse around best practises (at least I didn't find anything). So currently I have a big git repo with all playbooks and roles. I would like to automatically run a task whenever the correspondant role or playbook changed.
If I just use the internal cron check, it just checks for commits in the big git repo and runs if there was actually a change related to the playbook or another one.

Ideally I would like to:

- run playbook again when playbook file changed
- run role again when any related role file changed
- run role again when any related global variables changed

I was thinking about a git posthook script to detect the change and trigger the task run via api.

Alternativly split the big repo in a lot of small repos and use the internal cron check for every task.

Thank you for advice

**Comment by cf-sewe** [comment]
With the current implementation, you could already build something based on GitHub Action and Semaphore API:

1. Trigger GitHub action (assuming your code is in GitHub) on Commit / PR Merge
2. Within GitHub action call Semaphore [Tasks API](https://www.ansible-semaphore.com/api-docs/#/project/post_project__project_id__tasks) and start a new job.
  To identify which job to start, you may create a small config file and push it to the GIT, and evaluate that from the GitHub action. 

GitLab etc. should also be possible in similar fashion. 
Of course, Semaphore could also automatically check GIT for changes, however IMHO this over complicates  the implementation (thats what the API is for in my opinion).

**Reply by cdpb** [reply]
thank you for your answer. I am using a selfhosted gogs git server, which is capabile of pre and post hooks only. I could call an API from there too, that was my idea.

So you basically have a config file mapping the task template with a playbook and a role? GitHub checks that file and calls the API according to the changes?

**Reply by cf-sewe** [reply]
If you have a self-hosted server, I would simply create a small cron job that checks for fresh commits, or a nodejs/go daemon that receives the gogs webhook. That program would then initiate the Semaphore Job as required.

**Reply by cdpb** [reply]
yes, that was my idea too.
But as everything is in one big git repo, the script should not only run any task via api, it should be capable of detecting the change, down to the role, globale variable and playbook to run exactly the correct task(s).
Thats not terrible complicated of course, the general question is more about ready to go scripts or best practices.

If there is no such tool (other than maybe GitHub Actions) - I am happy to program it myself and I got my answer too. Thank you @cf-sewe

**Reply by gulbinas** [reply]
i think this can make it happen: https://github.com/ansible-semaphore/semaphore/discussions/1617
